File: pipeline-dashboard/server/scheduler.js
'use strict';

/**
 * Timezone-aware daily scheduler for the master-pipeline digest.
 *
 * Fires a callback at specific wall-clock times in named IANA timezones
 * (e.g. 09:00 Asia/Kolkata and 09:00 America/Los_Angeles), correctly handling
 * DST because it derives the target instant from the zone's *current* offset.
 *
 * No cron daemon and no third-party deps — pure setTimeout re-armed after each
 * fire. Ideal for the zero-dependency deployment model of this app.
 */

const { MASTER_DIGEST, SLACK } = require('./config');
const store = require('./store');
const { postMasterDigest, verifyAuth } = require('./slack');

/** Get the parts of "now" in a given IANA timezone as numbers. */
function nowInZone(tz) {
  const fmt = new Intl.DateTimeFormat('en-US', {
    timeZone: tz,
    hour12: false,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  });
  const parts = {};
  for (const p of fmt.formatToParts(new Date())) {
    if (p.type !== 'literal') parts[p.type] = Number(p.value);
  }
  // en-US formats midnight hour as 24; normalize to 0.
  if (parts.hour === 24) parts.hour = 0;
  return parts;
}

/**
 * Compute ms until the next occurrence of hour:minute in `tz`.
 * Works by measuring the offset between the zone's wall clock and UTC "now",
 * then projecting the target wall-clock instant into real (UTC) time.
 */
function msUntilNext(hour, minute, tz) {
  const now = new Date();
  const z = nowInZone(tz);

  // The zone's current wall-clock time as a UTC-epoch reference.
  const zoneNowUTC = Date.UTC(z.year, z.month - 1, z.day, z.hour, z.minute, z.second);
  // The zone's offset from UTC right now (ms). e.g. IST = +5:30.
  const offset = zoneNowUTC - now.getTime();

  // Target wall-clock time today in the zone, expressed as the same UTC-epoch basis.
  let targetZoneUTC = Date.UTC(z.year, z.month - 1, z.day, hour, minute, 0);
  // Convert that wall-clock target into a real instant by removing the offset.
  let targetReal = targetZoneUTC - offset;

  // If it already passed (or is within 1s), schedule for tomorrow.
  if (targetReal <= now.getTime() + 1000) {
    targetReal += 24 * 60 * 60 * 1000;
  }
  return targetReal - now.getTime();
}

async function fireDigest(label) {
  const failing = store.getMasterFailures(MASTER_DIGEST.failThreshold);
  if (!failing.length) {
    console.log(
      `[digest] (${label}) no master pipeline at >= ${MASTER_DIGEST.failThreshold} consecutive failures; nothing to post.`
    );
    return;
  }
  const outcome = await postMasterDigest(failing, {
    threshold: MASTER_DIGEST.failThreshold,
    when: label,
  });
  console.log(
    `[digest] (${label}) ${failing.length} failing master pipeline(s) -> post ${
      outcome.sent ? 'SENT' : `NOT sent (${outcome.reason})`
    } to ${MASTER_DIGEST.channel}`
  );
}

/** Arm one repeating daily timer for a single { hour, minute, tz } slot. */
function armSlot(slot) {
  const label = `${String(slot.hour).padStart(2, '0')}:${String(slot.minute).padStart(2, '0')} ${slot.tz}`;
  const schedule = () => {
    const delay = msUntilNext(slot.hour, slot.minute, slot.tz);
    const when = new Date(Date.now() + delay).toISOString();
    console.log(`[digest] next post for ${label} in ${(delay / 3600000).toFixed(2)}h (~${when})`);
    setTimeout(async () => {
      try {
        await fireDigest(label);
      } catch (e) {
        console.error(`[digest] (${label}) failed:`, e.message);
      } finally {
        schedule(); // re-arm for the next day
      }
    }, delay).unref?.();
  };
  schedule();
}

/** Start the master-digest scheduler. Safe no-op when disabled. */
function startMasterDigest() {
  if (!MASTER_DIGEST.enabled) {
    console.log('[digest] master digest disabled (MASTER_DIGEST_ENABLED=false).');
    return;
  }
  if (!SLACK.botToken) {
    console.warn(
      '[digest] SLACK_BOT_TOKEN not set — digest will run but only LOG (never post). Set the bot token to enable posting.'
    );
  } else {
    verifyAuth()
      .then((a) =>
        a.ok
          ? console.log(`[digest] Slack bot authenticated as ${a.user} (team ${a.team}).`)
          : console.error(`[digest] Slack auth.test failed: ${a.error}. Posting will fail.`)
      )
      .catch(() => {});
  }
  if (!MASTER_DIGEST.times.length) {
    console.warn('[digest] no MASTER_DIGEST_TIMES configured; scheduler idle.');
    return;
  }
  for (const slot of MASTER_DIGEST.times) armSlot(slot);
}

module.exports = { startMasterDigest, fireDigest, msUntilNext };

File: pipeline-dashboard/go/internal/scheduler/scheduler.go
// Package scheduler fires the master-pipeline digest at configured wall-clock
// times in named IANA timezones (DST-aware via time.LoadLocation).
// Ported from server/scheduler.js.
package scheduler

import (
	"fmt"
	"log"
	"time"

	"github.com/nutanix/msp-pipeline-dashboard/internal/config"
	"github.com/nutanix/msp-pipeline-dashboard/internal/slack"
	"github.com/nutanix/msp-pipeline-dashboard/internal/store"
)

// nextOccurrence returns the next time hour:minute occurs in tz (today or tomorrow).
func nextOccurrence(hour, minute int, tz string) time.Time {
	loc, err := time.LoadLocation(tz)
	if err != nil {
		log.Printf("[digest] unknown timezone %q, falling back to UTC: %v", tz, err)
		loc = time.UTC
	}
	now := time.Now().In(loc)
	target := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, loc)
	if !target.After(now.Add(time.Second)) {
		target = target.Add(24 * time.Hour)
	}
	return target
}

// FireDigest evaluates the threshold and posts if anything qualifies.
func FireDigest(label string) slack.Outcome {
	outcome := slack.Outcome{Sent: false, Skipped: true, Reason: "failure-digest-disabled"}
	if config.MasterDigest.Enabled {
		failing := store.GetMasterFailures(config.MasterDigest.FailThreshold)
		if len(failing) == 0 {
			log.Printf("[digest] (%s) no master pipeline at >= %d consecutive failures; nothing to post.",
				label, config.MasterDigest.FailThreshold)
			outcome = slack.Outcome{Sent: false, Skipped: true, Reason: "nothing-failing"}
		} else {
			outcome = slack.PostMasterDigest(failing, slack.DigestMeta{
				Threshold: config.MasterDigest.FailThreshold,
				When:      label,
			})
			status := "SENT"
			if !outcome.Sent {
				status = fmt.Sprintf("NOT sent (%s)", outcome.Reason)
			}
			log.Printf("[digest] (%s) %d failing master pipeline(s) -> post %s to %s",
				label, len(failing), status, config.MasterDigest.Channel)
		}
	}
	fireSuccessDigest(label)
	return outcome
}

func fireSuccessDigest(label string) {
	if !config.SuccessDigest.Enabled {
		return
	}
	ok := store.GetMasterSuccesses(config.SuccessDigest.Threshold)
	if len(ok) == 0 {
		log.Printf("[digest] (%s) no master pipeline at >= %d consecutive successes; success digest skipped.",
			label, config.SuccessDigest.Threshold)
		return
	}
	outcome := slack.PostSuccessDigest(ok, slack.DigestMeta{
		Threshold: config.SuccessDigest.Threshold,
		When:      label,
	})
	status := "SENT"
	if !outcome.Sent {
		status = fmt.Sprintf("NOT sent (%s)", outcome.Reason)
	}
	log.Printf("[digest] (%s) %d succeeding master pipeline(s) -> post %s to %s",
		label, len(ok), status, config.MasterDigest.Channel)
}

// FireDigestTest always attempts a Slack post so operators can verify the
// configured channel (SLACK_CHANNEL / MASTER_DIGEST_CHANNEL): the failure
// digest if anything qualifies, otherwise an all-clear. Mentions follow .env.
func FireDigestTest() slack.Outcome {
	failing := store.GetMasterFailures(config.MasterDigest.FailThreshold)
	meta := slack.DigestMeta{Threshold: config.MasterDigest.FailThreshold, When: "manual-test"}
	if len(failing) == 0 {
		outcome := slack.PostAllClear(meta)
		if outcome.Sent && outcome.Reason == "" {
			outcome.Reason = "all-clear"
		}
		status := "SENT"
		if !outcome.Sent {
			status = fmt.Sprintf("NOT sent (%s)", outcome.Reason)
		}
		log.Printf("[digest] (manual-test) all-clear -> post %s to %s", status, config.MasterDigest.Channel)
		fireSuccessDigest("manual-test")
		return outcome
	}
	return FireDigest("manual-test")
}

// armSlot schedules a repeating daily timer for one { hour, minute, tz } slot.
func armSlot(slot config.DigestTime) {
	label := fmt.Sprintf("%02d:%02d %s", slot.Hour, slot.Minute, slot.TZ)
	var schedule func()
	schedule = func() {
		next := nextOccurrence(slot.Hour, slot.Minute, slot.TZ)
		delay := time.Until(next)
		log.Printf("[digest] next post for %s in %.2fh (~%s)", label, delay.Hours(), next.UTC().Format(time.RFC3339))
		time.AfterFunc(delay, func() {
			func() {
				defer func() {
					if r := recover(); r != nil {
						log.Printf("[digest] (%s) failed: %v", label, r)
					}
				}()
				FireDigest(label)
			}()
			schedule() // re-arm for the next day
		})
	}
	schedule()
}

// Start launches the master-digest scheduler. Safe no-op when disabled.
func Start() {
	if !config.MasterDigest.Enabled && !config.SuccessDigest.Enabled {
		log.Printf("[digest] master + success digests disabled.")
		return
	}
	if !config.MasterDigest.Enabled {
		log.Printf("[digest] failure digest disabled (MASTER_DIGEST_ENABLED=false); success digest still scheduled.")
	}
	if !config.Slack.Enabled {
		log.Printf("[digest] Slack posting paused (SLACK_ENABLED=false). Digest will run but never post.")
	} else if config.Slack.BotToken == "" {
		log.Printf("[digest] SLACK_BOT_TOKEN not set — digest will run but only LOG (never post). Set the bot token to enable posting.")
	} else {
		a := slack.VerifyAuth()
		if a.OK {
			log.Printf("[digest] Slack bot authenticated as %s (team %s).", a.User, a.Team)
		} else {
			log.Printf("[digest] Slack auth.test failed: %s. Posting will fail.", a.Error)
		}
	}
	if len(config.MasterDigest.Times) == 0 {
		log.Printf("[digest] no MASTER_DIGEST_TIMES configured; scheduler idle.")
		return
	}
	for _, slot := range config.MasterDigest.Times {
		armSlot(slot)
	}
}
